hashintel/hash
# Fix crash on slow networks when creating entity type

## 🌟 What is the purpose of this PR?

The links table on an entity type is supposed to hide itself when the entity types are still loading, but at some point the context hook to check if loading was broken (probably by me) and would throw if called before entity types are loaded. This fixes that.

I've also taken the opportunity to rewrite the entity types context provider and use it in more places so we're not calling to the API repeatedly all over the place, and data updates everywhere in sync. 




## 🔗 Related links





- [Asana task](https://app.asana.com/0/1203179076056209/1203639679774982/f) _(internal)_

## 🔍 What does this change?




- Fixed crash by checking if link types are loading being rendering table properly instead of throwing when they're not
- Abstract an `EntityTypesContextProvider` (making use of the existing `useEntityTypesContextValue`) and place it at the top level instead of around the entity type editor. Rebuild file structure for this. 
- Change `useEntityTypesContextValue` so it doesn't fetch entity types in a session until the first time a component asks for them (`ensureFetched`) and to make the whole subgraph available. Also introduced an explicit `loading` flag
- New `optional` hooks for requesting entity types / link types from context, without throwing if they've not yet loaded. 
- Use entity type context in:
  - Entity type selector on new entity page (deleting the context it used)
  - Entity type editor (already used here for the entity type selector – but have rewritten the code for `useEntityTypeContextValue` to be based on the entity type context rather than fetching it itself)
  - Entity type list in sidebar

## ⚠️ Known issues



- When you publish an update to an entity type, the order of properties will change. This is because the version returned from the API is somewhat random. This will be fixed when [property types are sorted](https://app.asana.com/0/1203543021352032/1203661700052726/f) (_internal_)

## 🐾 Next steps



- [! Make property table sorted alphabetically by title](https://app.asana.com/0/1203543021352032/1203661700052726) (_internal_)
- [Potentially make use of a data store](https://hashintel.slack.com/archives/C022217GAHF/p1673361039956769) (_internal_)

## 🛡 What tests cover this?



-

## ❓ How to test this?



1.  Turn on network throttling
2. Create an entity type
3. Ensure no crash
